### PR TITLE
Reimplement Path, Contours and Polygons plots

### PR DESCRIPTION
**Pull request by philippjfr**
Over the past few months I've slowly been laying the groundwork to support more powerful Path, Contours and Polygons elements and in this PR I am also finally exposing this new functionality in plotting classes. This PR finishes off some final issues with the internally facing API and will reimplements the plotting classes based on this new API.

New features:

* Path, Contours and Polygons can now have any number of value dimensions associated with them which are revealed while hovering and which can be colored by.
* Each geometry in a Contours or Polygons element can now have a different value associated with it, so you no longer have to create huge NdOverlays with one element per geometry.
* Path now allows supplying varying value dimensions which are drawn as paths with continuously varying color.

[Here](https://anaconda.org/philippjfr/new_path_interface/notebook) is a demo of the new (and old) features.

 This PR will address the following issues:

* https://github.com/ioam/holoviews/issues/102
* https://github.com/ioam/holoviews/issues/416
* https://github.com/ioam/holoviews/issues/1558
* https://github.com/ioam/holoviews/issues/1963
* https://github.com/ioam/holoviews/issues/1856

and will allow me to finish off my geopandas interface in geoviews:

* https://github.com/ioam/geoviews/pull/88

### CODE REVIEW

**Comment by jbednar** [comment]
Fabulous! Was the county name meant to be visible in the hover?

image

**Comment by philippjfr** [comment]
>Fabulous! Was the county name meant to be visible in the hover?

Yes, not quite done yet :-)

**Comment by philippjfr** [comment]
Actually nevermind, it does work, I just hadn't declared County as a dimension.

**Comment by philippjfr** [comment]
@jlstevens This is ready for review once you have time.

**Comment by philippjfr** [comment]
Additional user API is not part of this PR btw, the two main things I think we should still add are:

* Allow groupby by value dimensions on Contours/Polygon (but not Path)
* Allow selecting subpaths by index and value dimensions

**Comment by jlstevens** [comment]
It looks like you've found a few more things to improve in this PR. I'll review it once you think it is ready again...

**Comment by philippjfr** [comment]
>It looks like you've found a few more things to improve in this PR. I'll review it once you think it is ready again...

It's ready.

File: doc/Tutorials/Introduction.ipynb [hunk 1]
```diff
@@ -662,8 +661,8 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "green05 = levels.Overlay.Green[0.5]\n",
-    "green05 + green05.Channel + green05.Channel.Green.sample(y=0.0)"
+    "green05 = levels.Contours.Green\n",
+    "green05 + chans.Channel.Green + chans.Channel.Green.sample(y=0.0)"
```
Comment (jlstevens) [inline doc/Tutorials/Introduction.ipynb]: Doesn't this have implications for backwards compatibility?

File: examples/gallery/demos/bokeh/texas_choropleth_example.ipynb [hunk 2]
```diff
@@ -38,20 +38,11 @@
...
-    "county_ys = [county[\"lats\"] for county in counties.values()]\n",
-    "\n",
-    "county_names = [county['name'] for county in counties.values()]\n",
-    "county_rates = [unemployment[county_id] for county_id in counties]\n",
-    "\n",
-    "county_polys = {name: hv.Polygons((xs, ys), level=rate, vdims=['Unemployment'])\n",
-    "                for name, xs, ys, rate in zip(county_names, county_xs, county_ys, county_rates)}\n",
-    "\n",
-    "choropleth = hv.NdOverlay(county_polys, kdims=['County'])"
+    "choropleth = hv.Polygons(counties, ['lons', 'lats'], [('detailed name', 'County'), 'Unemployment'])"
```
Comment (jlstevens) [inline examples/gallery/demos/bokeh/texas_choropleth_example.ipynb]: I assume the old code still works and the new approach is just more succinct/flexible/efficient.

File: examples/reference/elements/bokeh/Polygons.ipynb [hunk 3]
```diff
@@ -28,9 +28,11 @@
...
+    "A ``Polygons`` represents a contiguous filled area in a 2D space as a list of paths. Just like the ``Contours`` element additional scalar value dimensions maybe may be supplied, which can be used to color the ``Polygons`` with the defined ``cmap``. Like other ``Path`` types it accepts a list of arrays, dataframes, a dictionary of columns (or any of the other literal formats including tuples of columns and lists of tuples).\n",
     "\n",
-    "Polygons with values can be used to build heatmaps with arbitrary shapes."
+    "In order to efficiently represent the scalar values associated with each path the dictionary format is preferable since it can store the scalar values without expanding them into a whole column. Additionally it allows passing multiple columns as a single array by specifying the dimension names as a tuple.\n",
+    "\n",
+    "In this example we will createa list of random polygons each with an associated ``level`` value. Polygons will default to using the first value dimension as the ``color_index`` but for clarity we will define the ``color_index`` explicitly:"
```
Comment (jlstevens) [inline examples/reference/elements/bokeh/Polygons.ipynb]: createa -> create a

File: examples/user_guide/Exporting_and_Archiving.ipynb [hunk 4]
```diff
@@ -155,7 +155,7 @@
    "source": [
     "%%opts Contours (linewidth=1.3) Image (cmap=\"gray\")\n",
     "cs = contours(penguins[:,:,'R'], levels=[0.10,0.80])\n",
-    "cs"
+    "penguins[:, :, 'R'] * cs"
```
Comment (jlstevens) [inline examples/user_guide/Exporting_and_Archiving.ipynb]: The operation no longer just adds contours on top of the input and return it? That's fine as long as there is a switch to restore the old behavior. I also suspect we may have made this change a while back.

File: holoviews/core/data/interface.py [hunk 5]
```diff
@@ -86,8 +86,12 @@ class Interface(param.Parameterized):
 
     datatype = None
 
+    # Denotes whether the interface expects gridded data
     gridded = False
 
+    # Denotes whether the interface expects multiple ragged arrays
```
Comment (jlstevens) [inline holoviews/core/data/interface.py]: Does it have to be arrays? Isn't it ragged 'data' (i.e multiple elements of different lengths)?

File: holoviews/operation/element.py [hunk 6]
```diff
@@ -415,22 +415,17 @@ class contours(Operation):
     filled = param.Boolean(default=False, doc="""
         Whether to generate filled contours""")
 
-    overlaid = param.Boolean(default=True, doc="""
+    overlaid = param.Boolean(default=False, doc="""
```
Comment (jlstevens) [inline holoviews/operation/element.py]: Ok, this answers my earlier question.

File: holoviews/plotting/bokeh/path.py [hunk 7]
```diff
@@ -6,25 +6,32 @@
...
+from .element import ColorbarPlot, line_properties, fill_properties
 from .util import expand_batched_style
 
 
-class PathPlot(ElementPlot):
+class PathPlot(ColorbarPlot):
+
+    color_index = param.ClassSelector(default=None, class_=(util.basestring, int),
+                                      allow_None=True, doc="""
+      Index of the dimension from which the color will the drawn""")
```
Comment (jlstevens) [inline holoviews/plotting/bokeh/path.py]: Would be nice to get the ``op`` proposal in soon...

File: tests/testoperation.py [hunk 8]
```diff
@@ -56,21 +56,21 @@ def test_image_gradient(self):
...
-        op_contours = contours(img)
-        ndoverlay = NdOverlay(None, kdims=['Levels'])
-        ndoverlay[0.5] = Contours([[(-0.5,  0.416667), (-0.25, 0.5)], [(0.25, 0.5), (0.5, 0.45)]],
-                                  group='Level', level=0.5, vdims=img.vdims)
-        self.assertEqual(op_contours, img*ndoverlay)
+        op_contours = contours(img, levels=[0.5])
+        contour = Contours([[(-0.5,  0.416667, 0.5), (-0.25, 0.5, 0.5)],
+                             [(0.25, 0.5, 0.5), (0.5, 0.45, 0.5)]],
+                            vdims=img.vdims)
+        print(contour.range(2))
```
Comment (jlstevens) [inline tests/testoperation.py]: Stray print statement.

File: examples/user_guide/Exporting_and_Archiving.ipynb (hunk 4 above)
Comment (jlstevens) [reply examples/user_guide/Exporting_and_Archiving.ipynb]: I see you changed the default - see comment below.

Comment (jlstevens) [reply examples/user_guide/Exporting_and_Archiving.ipynb]: We would at least need to document this change of behavior in the changelog/release notes.

File: holoviews/core/data/multipath.py [hunk 9]
```diff
@@ -20,7 +20,9 @@ class MultiInterface(Interface):
 
     datatype = 'multitabular'
 
-    subtypes = ['dataframe', 'dictionary', 'array', 'dask']
+    subtypes = ['dictionary', 'dataframe', 'array', 'dask']
```
Comment (jlstevens) [inline holoviews/core/data/multipath.py]: I *think* it makes sense to try the more general dictionary (i.e standard python literals) format first. Might be other implications I haven't figured out yet. Then again, ``MultiInterface`` is pretty new so it probably doesn't matter wrt backwards compatibility.

**Comment by jlstevens** [comment]
Looks good. My main comments are:

* I am assuming old code still works and that many of the updates to the notebooks just show the new, cleaner way of doing things.
* We need to be clear about any changes to behavior, e.g the way the contour function no longer overlays thing.
* The change in how you use attribute access in the introduction notebook seems like the most worrying change I noticed...
* Reading notebook diffs is tricky - I can't tell if you only updated the element notebooks or added material to showcase all the new functionality. Github's notebook rendering seems [off](https://github.com/ioam/holoviews/blob/bf217c5c8e762dde721a245851b95241ed556288/examples/reference/elements/bokeh/Contours.ipynb) (a problem with the HTML maybe?):

image

If you haven't added material it would be good to see more examples (e.g of the split method).

Anyway, none of this is critical and once a few of my comments above are addressed (typos, stray print and other minor things), I'm happy to merge. The main other thing I would like is more examples for the docs.

File: doc/Tutorials/Introduction.ipynb (hunk 1 above)
Comment (philippjfr) [reply doc/Tutorials/Introduction.ipynb]: Guess it has some, it's no longer ``overlaid`` by default. I could have made it be overlaid again but I generally don't want to.

File: examples/gallery/demos/bokeh/texas_choropleth_example.ipynb (hunk 2 above)
Comment (philippjfr) [reply examples/gallery/demos/bokeh/texas_choropleth_example.ipynb]: Right.

File: examples/reference/elements/bokeh/Polygons.ipynb (hunk 3 above)
Comment (philippjfr) [reply examples/reference/elements/bokeh/Polygons.ipynb]: Thanks.

File: examples/user_guide/Exporting_and_Archiving.ipynb (hunk 4 above)
Comment (philippjfr) [reply examples/user_guide/Exporting_and_Archiving.ipynb]: Right, may start requiring committers adding backwards compatibility changes to the CHANGELOG when they are made so we don't forget about them. Will do that here.

File: holoviews/core/data/interface.py (hunk 5 above)
Comment (philippjfr) [reply holoviews/core/data/interface.py]: True.

File: holoviews/plotting/bokeh/path.py (hunk 7 above)
Comment (philippjfr) [reply holoviews/plotting/bokeh/path.py]: Agreed.

File: tests/testoperation.py (hunk 8 above)
Comment (philippjfr) [reply tests/testoperation.py]: Thanks I'll delete it.

File: holoviews/core/data/multipath.py (hunk 9 above)
Comment (philippjfr) [reply holoviews/core/data/multipath.py]: Right, the point here is that it doesn't accidentally expand your scalar values into a dataframe column for instance.

**Comment by philippjfr** [comment]
>I am assuming old code still works and that many of the updates to the notebooks just show the new, cleaner way of doing things.

Correct.

>We need to be clear about any changes to behavior, e.g the way the contour function no longer overlays thing.

Yes, ``contours`` needs a note about backward compatibility.

>The change in how you use attribute access in the introduction notebook seems like the most worrying change I noticed...

This is a result of two things 1) it's no longer overlaid by default and b) contours now returns a single ``Contours`` element not an NdOverlay of them so this falls under "noting contours backward compatibility changes".

>Reading notebook diffs is tricky - I can't tell if you only updated the element notebooks or added material to showcase all the new functionality. Github's notebook rendering seems off (a problem with the HTML maybe?):

Not sure what you're referring to, looks fine to me.

>If you haven't added material it would be good to see more examples (e.g of the split method).

I'm working on a "Working with Geometries" user guide which will cover stuff like this. Will see how far I get with that, but I can add it in a later PR.

File: holoviews/core/data/multipath.py (hunk 9 above)
Comment (jlstevens) [reply holoviews/core/data/multipath.py]: Makes sense.

**Comment by jlstevens** [comment]
Ok great! 

Should contours get a compatibility flag in config? If we find ourselves with a number of small backwards incompatibility issues, we could group them under a new config flag.

That doesn't need to be decided right away and I'll merge once you get rid of that stray print and fix that typo.

**Comment by jlstevens** [comment]
The tests have now passed except for on transient error (build restarted and still running).

Merging.

**Pull request merged**